File: frontend/lib/models/category.dart
import 'dart:async';

import 'package:tracko/models/transaction.dart';
import 'package:tracko/models/user.dart';

class Category {
  int? id;
  String name = '';
  int? userId;
  String categoryType = 'EXPENSE';
  List<Transaction> transactions = [];

  Category();

  Category.make(this.id, this.name, this.userId);

  @override
  String toString() {
    return 'Category{id: $id, name: $name, userId: $userId, categoryType: $categoryType, transactions: $transactions}';
  }

  static int defaultCategoryId() {
    return 1;
  }
}

File: backend/src/main/java/com/trako/entities/Category.java
package com.trako.entities;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.trako.enums.CategoryType;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import lombok.Getter;
import lombok.Setter;

@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
@Entity
@Table(name = "categories")
@Getter
@Setter
public class Category {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Long id;

    @NotNull
    @Column(name = "name", length = 250)
    private String name;

    @NotNull
    @Column(name = "user_id", length = 36)
    private String userId;

    @Column(name = "is_roll_over_enabled")
    private Boolean isRollOverEnabled;

    @Column(name = "parent_category_id")
    private Long parentCategoryId;

    @NotNull
    @Column(name = "category_type", length = 16)
    @Enumerated(EnumType.STRING)
    private CategoryType categoryType = CategoryType.EXPENSE;

    @JsonIgnore
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", insertable = false, updatable = false)
    private User user;

    @JsonIgnore
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_category_id", insertable = false, updatable = false)
    private Category parentCategory;
}
